Interop/ModelCard/src/main/java/ModelCard.java

```diff
@@ -0,0 +1,74 @@
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.tribuo.Model;
+import org.tribuo.evaluation.Evaluation;
+import org.tribuo.interop.ExternalModel;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Paths;
+
+public class ModelCard {
+    private static final ObjectMapper mapper = new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT);
+    private final ModelDetails modelDetails;
+    private final TrainingDetails trainingDetails;
+    private final TestingDetails testingDetails;
+    private final UsageDetails usageDetails;
+
+    public ModelCard(Model<?> model, Evaluation<?> evaluation) {
+        if (model instanceof ExternalModel) throw new IllegalArgumentException();
+        modelDetails = new ModelDetails(model);
+        trainingDetails = new TrainingDetails(model);
+        testingDetails = new TestingDetails(evaluation);
+        usageDetails = new UsageDetails();
+    }
+
+    public ModelCard(String sourceFile) throws IOException {
+        JsonNode modelCardJson = mapper.readTree(Paths.get(sourceFile).toFile());
+        modelDetails = new ModelDetails(modelCardJson.get("ModelDetails"));
+        trainingDetails = new TrainingDetails(modelCardJson.get("TrainingDetails"));
+        testingDetails = new TestingDetails(modelCardJson.get("TestingDetails"));
+        usageDetails = new UsageDetails(modelCardJson.get("UsageDetails"));
+    }
+
+    public ModelDetails getModelDetails() {
+        return modelDetails;
+    }
+
+    public TrainingDetails getTrainingDetails() {
+        return trainingDetails;
+    }
+
+    public TestingDetails getTestingDetails() {
+        return testingDetails;
+    }
+
+    public UsageDetails getUsageDetails() {
+        return usageDetails;
+    }
+
+    public void addMetric(String metricDescription, Double metricValue) {
+        testingDetails.addMetric(metricDescription, metricValue);
+    }
+
+    public ObjectNode toJson() {
+        ObjectNode modelCardObject = mapper.createObjectNode();
+        modelCardObject.set("ModelDetails", modelDetails.toJson());
+        modelCardObject.set("TrainingDetails", trainingDetails.toJson());
+        modelCardObject.set("TestingDetails", testingDetails.toJson());
+        modelCardObject.set("UsageDetails", usageDetails.toJson());
+        return modelCardObject;
+    }
+
+    public void saveToFile(String destinationFile) throws IOException {
+        ObjectNode modelCardObject = toJson();
+        mapper.writeValue(new File(destinationFile), modelCardObject);
+    }
+
+    @Override
+    public String toString() {
```

Comment:
These classes should all implement `equals` (and consequently `hashCode`) too. You should be able to autogenerate them in IDEs like IntelliJ.

Reply:
Done. Right now, for the implementations of `equals` in `ModelDetails`, I'm comparing the text value of the `JsonNode configuredParams` fields for equality rather than directly comparing the objects. This is because in my first constructor for `ModelDetails`, `configuredParams` has an actual type of `ObjectNode` whereas in the second constructor (when building a ModelCard object from a file), the actual type is `JsonNode`. I am aware that the two are used rather interchangeably in practice and in the context that I'm using them in, but I wanted to run this past someone to make sure that it would be okay to do this and just compare their text values in my implementation of `equals`.